ddtrace/contrib/grpc/patch.py

```diff
@@ -1,64 +1,105 @@
+from ddtrace import config, Pin
 
-from .client_interceptor import GrpcClientInterceptor
+from ...ext import AppTypes
+from ...utils.wrappers import unwrap as _u
+
+from .client_interceptor import create_client_interceptor
+from .server_interceptor import create_server_interceptor
+
+
+config._add('grpc', dict(
+    service_name='grpc',
+    app='grpc',
+    span_type='grpc',
+    app_type=AppTypes.web,
+
+    distributed_tracing_enabled=True,
+))
 
 
 def patch():
     # patch only once
     if getattr(grpc, '__datadog_patch', False):
         return
     setattr(grpc, '__datadog_patch', True)
-    Pin(service='grpc', app='grpc', app_type='grpc').onto(grpc)
 
-    _w = wrapt.wrap_function_wrapper
+    Pin(service=config.grpc.service_name,
+        app=config.grpc.app,
+        app_type=config.grpc.app_type).onto(grpc)
 
-    _w('grpc', 'insecure_channel', _insecure_channel_with_interceptor)
-    _w('grpc', 'secure_channel', _secure_channel_with_interceptor)
+    _w('grpc', 'insecure_channel', _client_channel_interceptor)
+    _w('grpc', 'secure_channel', _client_channel_interceptor)
+    _w('grpc', 'server', _server_constructor_interceptor)
 
 
 def unpatch():
     if not getattr(grpc, '__datadog_patch', False):
         return
     setattr(grpc, '__datadog_patch', False)
-    unwrap(grpc, 'secure_channel')
-    unwrap(grpc, 'insecure_channel')
 
+    pin = Pin.get_from(grpc)
+    if pin:
+        pin.remove_from(grpc)
 
-def _insecure_channel_with_interceptor(wrapped, instance, args, kwargs):
+    _u(grpc, 'secure_channel')
+    _u(grpc, 'insecure_channel')
+    _u(grpc, 'server')
+
+
+def _client_channel_interceptor(wrapped, instance, args, kwargs):
     channel = wrapped(*args, **kwargs)
 
-    if 'target' in kwargs:
-        target = kwargs['target']
-    else:
-        target = args[0]
+    (host, port) = _parse_target_from_arguments(args, kwargs)
+
+    # DEV: we clone the pin on the grpc module and configure it for the client
+    # interceptor
+    pin = Pin.get_from(grpc)
+    if not pin:
+        return channel
+
+    tags = {
+        'grpc.host': host,
+        'grpc.port': port,
+    }
+    if pin and pin.tags:
+        tags.update(pin.tags)
+
+    pin = pin.clone(tags=tags)
+
+    channel = grpc.intercept_channel(channel, create_client_interceptor(pin))
 
-    (host, port) = get_host_port(target)
-    channel = _intercept_channel(channel, host, port)
     return channel
 
 
-def _secure_channel_with_interceptor(wrapped, instance, args, kwargs):
-    channel = wrapped(*args, **kwargs)
+def _server_constructor_interceptor(wrapped, instance, args, kwargs):
+    # DEV: we clone the pin on the grpc module and configure it for the server
+    # interceptor
 
-    if 'target' in kwargs:
-        target = kwargs['target']
+    pin = Pin.get_from(grpc)
+    if not pin:
+        return wrapped(*args, **kwargs)
+
+    interceptor = create_server_interceptor(pin)
+
+    if 'interceptors' in kwargs:
```

Comment:
This probably deserves a comment, like `Inject our tracing interceptor first in the list of interceptors` or something.